### PR TITLE
8299852: Modernize ConcurrentHashMap

### PR DESCRIPTION
**Pull request by minborg**
`java.util.concurrent.ConcurrentHashMap` is relatively old and has not been updated to reflect the current state of Java and can be modernized: 

* Add `@Serial` annotations 
* Seal classes and restrict subclassing for internal classes 
* Use pattern matching for instance 
* Remove redundant modifiers (such as some final declarations) 
* Use Objects.requireNonNull for invariant checking 
* Use diamond operators instead of explicit types 
* Simplify expressions using Math::max

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299852](https://bugs.openjdk.org/browse/JDK-8299852): Modernize ConcurrentHashMap

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/11924/head:pull/11924` \
`$ git checkout pull/11924`

Update a local copy of the PR: \
`$ git checkout pull/11924` \
`$ git pull https://git.openjdk.org/jdk.git pull/11924/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11924`

View PR using the GUI difftool: \
`$ git pr show -t 11924`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11924.diff">https://git.openjdk.org/jdk/pull/11924.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/11924#issuecomment-1377056880)

### CODE REVIEW

**Comment by pavelrappo** [comment]
I'm sure others will note or have already noted elsewhere that we don't do direct PRs against JSR 166.

**Comment by minborg** [comment]
> I'm sure others will note or have already noted elsewhere that we don't do direct PRs against JSR 166.

Yes. I will change the PR to a draft as we explore ways to integrate the spirit of the proposed changes.

**Comment by Martin-Buchholz** [comment]
ObMention @DougLea

File: src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java [hunk 1]
```diff
@@ -99,7 +101,7 @@
  * Bear in mind that the results of aggregate status methods including
  * {@code size}, {@code isEmpty}, and {@code containsValue} are typically
  * useful only when a map is not undergoing concurrent updates in other threads.
- * Otherwise the results of these methods reflect transient states
+ * Otherwise, the results of these methods reflect transient states
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java:104]: I agree the added comma here is an improvement.

File: src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java [hunk 2]
```diff
@@ -351,7 +354,7 @@
      * Actual hash code distributions encountered in practice
      * sometimes deviate significantly from uniform randomness.  This
      * includes the case when N > (1<<30), so some keys MUST collide.
-     * Similarly for dumb or hostile usages in which multiple keys are
+     * Similarly, for dumb or hostile usages in which multiple keys are
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java:357]: "Similarly" should almost always be followed by a comma, but this is an exception.

File: src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java [hunk 3]
```diff
@@ -622,7 +626,7 @@
      * are special, and contain null keys and values (but are never
      * exported).  Otherwise, keys and vals are never null.
      */
-    static class Node<K,V> implements Map.Entry<K,V> {
+    static sealed class Node<K,V> implements Map.Entry<K,V> {
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java:629]: There is a long term modernization task of making nested classes private to make use of
https://openjdk.org/jeps/181
It looks like this is an example.

File: src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java [hunk 4]
```diff
@@ -1999,7 +1998,7 @@ else if (f instanceof ReservationNode)
             }
         }
         if (delta != 0)
-            addCount((long)delta, binCount);
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java:2002]: Not sure why delta is not declared long as in clear().  But if not, then the explicit cast is documentation of the implementation choice to declare it int.

**Comment by ExE-Boss** [comment]
This PR should probably also add the [`@Override`] annotation to overridden methods in the [`ConcurrentHashMap`] class.

[`@Override`]: https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Override.html
[`ConcurrentHashMap`]: https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html

File: src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java [hunk 5]
```diff
@@ -974,12 +987,12 @@ public boolean containsKey(Object key) {
...
      */
+    @Override
     public boolean containsValue(Object value) {
-        if (value == null)
-            throw new NullPointerException();
+        Objects.requireNonNull(value);
         Node<K,V>[] t;
         if ((t = table) != null) {
-            Traverser<K,V> it = new Traverser<K,V>(t, t.length, 0, t.length);
+            Traverser<K,V> it = new Traverser<>(t, t.length, 0, t.length);
```
Comment (szegedi) [inline src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java:995]: Using diamond is one way to modernize this code, using `var` would be another, even more concise way. Consider: 
```suggestion
            var it = new Traverser<K, V>(t, t.length, 0, t.length);
```
I don't support blindly using `var` everywhere, but I believe when the RHS of the assignment spells out the type that's exactly the kind of usage envisioned for it.

File: src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java [hunk 6]
```diff
@@ -2887,8 +2887,9 @@ private void unlockRoot() {
         /**
          * Possibly blocks awaiting root lock.
          */
+
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java:2890]:
```suggestion
```

**Comment by minborg** [comment]
I think we should close this PR and maybe come back at a later stage with these proposed improvements.

**Pull request closed without merging**